input/elasticapm/internal/modeldecoder/modeldecoderutil/http.go

```diff
@@ -45,17 +45,18 @@ func HTTPHeadersToStructPb(h http.Header) *structpb.Struct {
 	return nil
 }
 
-func HTTPHeadersToModelpb(h http.Header) map[string]*modelpb.HTTPHeaderValue {
+func HTTPHeadersToModelpb(h http.Header) []*modelpb.HTTPHeader {
 	if len(h) == 0 {
 		return nil
 	}
-	m := make(map[string]*modelpb.HTTPHeaderValue, len(h))
+	headers := make([]*modelpb.HTTPHeader, 0, len(h))
 	for k, v := range h {
-		m[k] = &modelpb.HTTPHeaderValue{
-			Values: v,
-		}
+		headers = append(headers, &modelpb.HTTPHeader{
+			Key:   k,
+			Value: v,
```

Comment:
Should we be copying the slice here? Previously we were cloning the `http.Header` value when translating, so any mutation of the slice wouldn't impact the original value. I think this matters since the decoder reuses memory.

Reply:
I might be missing something but isn't the object reset before putting it back into the pool ? The original `http.Header` is cleared and the slice never mutated. When could it be mutated ?

Reply:
Any model processor _could_ modify the slice. They might not now, but we need to be defensive here since the input model is pooled and could be reused across two different connections.

Reply:
Sorry, I didn’t understand. Can you clarify:

Suppose we are mapping the `errorEvent`, the input model is fetched from the pool, mapped to modelpb and released into the pool: https://github.com/elastic/apm-data/blob/c6848e06bed4f3cd116ed203b13bcf4e4cd3887b/input/elasticapm/internal/modeldecoder/v2/decoder.go#L173-L174

Before being released, the object is reset: https://github.com/elastic/apm-data/blob/c6848e06bed4f3cd116ed203b13bcf4e4cd3887b/input/elasticapm/internal/modeldecoder/v2/decoder.go#L105

The reset call is recursive and it's gonna clear the `HTTPHeader`: https://github.com/elastic/apm-data/blob/c6848e06bed4f3cd116ed203b13bcf4e4cd3887b/input/elasticapm/internal/modeldecoder/nullable/nullable.go#L316

My understanding is that the original `http.Header` is cleared and all the slices it contains are never mutated. Once the original object is reset and the http header is cleared, the only reference to the "original values" (the slices) is the one we passed to the `modelpb`.

The model processors would then be able to modify the `modelpb` events which contains slices that are no longer referenced by the input models.

Reply:
I was thinking that the `http.Header` was pooled as well, but it's not - a new one is created in the nullable.HTTPHeader json-iterator callback. So this should be safe after all.